torch/quantization/observer.py

```diff
@@ -137,16 +144,80 @@ def calculate_qparams(self):
 
     @torch.jit.export
     def extra_repr(self):
-        return 'min_val={}, max_val={}'.format(self.min_val, self.max_val)
+        return "min_val={}, max_val={}".format(self.min_val, self.max_val)
+
+
+class HistogramObserver(ObserverBase):
+    r"""
+    The module records the running histogram of tensor values along with
+    min/max values. calculate_qparams will calculate scale and zero_point
+    """
+
+    __annotations__ = {
+        "min_val": Optional[torch.Tensor],
+        "max_val": Optional[torch.Tensor],
+        "relaxed_min": torch.Tensor,
+        "relaxed_max": torch.Tensor,
+        "histogram": Optional[torch.Tensor],
+    }
+
+    def __init__(self, bins=2048, **kwargs):
+        super(HistogramObserver, self).__init__(**kwargs)
+        self.bins = bins
+        self.histogram = None
+        self.min_val = None
+        self.max_val = None
+        self.relaxed_min = torch.tensor([0])
+        self.relaxed_max = torch.tensor([0])
+
+    def forward(self, x):
+        min_val = self.min_val
+        max_val = self.max_val
+        histogram = self.histogram
+        if min_val is None or max_val is None or histogram is None:
+            min_val = torch.min(x)
+            max_val = torch.max(x)
+            range = max_val - min_val
+            self.relaxed_min = min_val - 0.5 * range
+            self.relaxed_max = max_val + 0.5 * range
+            self.histogram = torch.histc(x, self.bins, min=self.relaxed_min.item(), max=self.relaxed_max.item())
+            self.min_val = self.relaxed_min
+            self.max_val = self.relaxed_max
+        else:
+            new_histogram = torch.histc(x, self.bins, min=self.relaxed_min.item(), max=self.relaxed_max.item())
```

Comment:
Please add a warning if the min/max of the incoming data in subsequent batches is outside the relaxed_min/max range.